python/ray/ml/examples/pytorch/permuted_mnist_domain_incremental_learning_example.py

```diff
@@ -0,0 +1,297 @@
+
+    """
+
+    def __init__(self, n_tasks: int = 3):
+        self.n_tasks = n_tasks
+        self.permutations = [
+            np.random.permutation(28 * 28) for _ in range(self.n_tasks)
+        ]
+
+    def random_permute_dataset(
+        self, dataset: ray.data.Dataset, permutation: np.ndarray
+    ):
+        """Randomly permutes the pixels for each image in the dataset."""
+
+        # permutation transformation
+        class PixelsPermutation(object):
+            def __call__(self, batch):
+                permuted_batch = [
+                    (image.reshape(-1)[permutation].reshape(28, 28), label)
+                    for (image, label) in batch
+                ]
+                return permuted_batch
+
+        return dataset.map_batches(PixelsPermutation, compute=ActorPoolStrategy())
+
+    def generate_train_stream(self) -> Iterator[ray.data.Dataset]:
+        for permutation in self.permutations:
+            mnist_dataset = get_mnist_dataset(train=True)
+            permuted_mnist_dataset = self.random_permute_dataset(
+                mnist_dataset, permutation
+            )
+            yield convert_dataset_to_pandas_format(permuted_mnist_dataset)
+
+    def generate_test_stream(self) -> Iterator[ray.data.Dataset]:
+        for permutation in self.permutations:
+            mnist_dataset = get_mnist_dataset(train=False)
+            permuted_mnist_dataset = self.random_permute_dataset(
+                mnist_dataset, permutation
+            )
+            yield convert_dataset_to_pandas_format(permuted_mnist_dataset)
+
+
+# TODO: Handle this automatically in TorchVisionDatasource
+def convert_dataset_to_pandas_format(image_dataset: ray.data.Dataset):
+    """Converts the image dataset consisting of tuples to Pandas format.
+
+    Converting to Pandas format is necessary so that Dataset can be used with
+    Preprocessors and Predictors.
+    """
+
+    def convert_batch_to_pandas(batch):
+        images = [item[0] for item in batch]
+        labels = [item[1] for item in batch]
+
+        df = pd.DataFrame({"image": TensorArray(images), "label": labels})
+
+        return df
+
+    return image_dataset.map_batches(convert_batch_to_pandas)
+
+
+def train_loop_per_worker(config: dict):
+    num_epochs = config["num_epochs"]
+    learning_rate = config["learning_rate"]
+    momentum = config["momentum"]
+    batch_size = config["batch_size"]
+
+    model = SimpleMLP(num_classes=10)
+
+    # Load model from checkpoint if there is a checkpoint to load from.
+    checkpoint_to_load = train.load_checkpoint()
+    if checkpoint_to_load:
+        state_dict_to_resume_from = checkpoint_to_load["model"]
+        model.load_state_dict(state_dict=state_dict_to_resume_from)
+
+    model = train.torch.prepare_model(model)
+
+    optimizer = SGD(model.parameters(), lr=learning_rate, momentum=momentum)
+    criterion = CrossEntropyLoss()
+
+    dataset_shard = train.get_dataset_shard("train").to_torch(
```

Comment:
for my education, at this point the "train" dataset is already sharded?

Reply:
Yes!